ring/partition_ring.go

```diff
@@ -75,11 +85,227 @@ func (r *PartitionRing) ActivePartitionForKey(key uint32) (int32, error) {
 	return 0, ErrNoActivePartitionFound
 }
 
+// ShuffleShard returns a subring for the provided identifier (eg. a tenant ID)
+// and size (number of partitions).
+//
+// The algorithm used to build the subring is a shuffle sharder based on probabilistic
+// hashing. We pick N unique partitions, walking the ring starting from random but
+// predictable numbers. The random generator is initialised with a seed based on the
+// provided identifier.
+//
+// This function returns a subring containing ONLY ACTIVE partitions.
+//
+// This function supports caching.
+//
+// This implementation guarantees:
+//
+//   - Stability: given the same ring, two invocations returns the same result.
+//
+//   - Consistency: adding/removing 1 partition from the ring generates a resulting
+//     subring with no more then 1 difference.
+//
+//   - Shuffling: probabilistically, for a large enough cluster each identifier gets a different
+//     set of instances, with a reduced number of overlapping instances between two identifiers.
+func (r *PartitionRing) ShuffleShard(identifier string, size int) (*PartitionRing, error) {
+	if cached := r.shuffleShardCache.getSubring(identifier, size); cached != nil {
+		return cached, nil
+	}
+
+	subring, err := r.shuffleShard(identifier, size, 0, time.Now())
```

Comment:
nit: do we need "time.Now()" when not using lookback?

Reply:
Agree. Done in https://github.com/grafana/dskit/pull/476/commits/19af036f63181298290f54ce0d30d1544acedc63